### PR TITLE
Optimize cuda scatter() on 2D compatible.

### PR DESCRIPTION
**Pull request by zhanghuanrong**
Faster rcnn model heavily use this case.

### CODE REVIEW

**Comment by ybrnathan** [comment]
@Duli is added to the review. #Closed

**Comment by ybrnathan** [comment]
@YulongW is added to the review. #Closed

File: onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu [hunk 1]
```diff
@@ -37,61 +64,162 @@ __global__ void _ScatterElementsKernel(
   output_data[data_idx] = updates[indices_index];
 }
 
+// From the innermost axis (largest) check equality of dim value of input and indices.
+// If same, merge it and continue. Otherwise, copy remaining. The scatter axis need
+// to be keep.
+static int CompactInputIndicesDims(
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu]: For <=2D, this should have a very quick return.

File: onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu [hunk 2]
```diff
@@ -7,6 +7,33 @@
...
+    const int64_t indices_size,
+    const fast_divmod indices_stride_row,
+    const T* updates,
+    const int64_t output_row_size,
+    T* output_data) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(indices_index, indices_size);
+
+  int row, col, data_idx;
+  indices_stride_row.divmod(indices_index, row, col);
+  int dim = (int)(indices_data[indices_index]);
```
Comment (fs-eire) [inline onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu]: is there potential overflow here?

Comment (zhanghuanrong) [reply onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu]: Agree. In general those use fast_divmod do not support large size. This is common bad in many current implementation. Yet we may force use gsl:narrow_cast check later for all possible space.

File: onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu (hunk 1 above)
Comment (zhanghuanrong) [reply onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu]: agree. As impact is low, will change it in future.

**Pull request merged**